### ISSUE TITLE
Cant mount a cd image for a game and install and run?

### ISSUE THREAD

**Issue by Blad3forc3** [ruanformigoni/gameimage#3]
This would work great if you could mount a cd image for an old windows game and install and run from within the appimage. Could this be done?

**Comment by ruanformigoni** [comment]
Hi @Blad3forc3 , you can mount the iso into the rom folder and it should work out of the box.

Option 1: You can use fuse iso, install it with:

```
sudo apt-get install fuseiso # for debian-based
sudo pacman -S fuseiso # for arch-base
```

And use it as `fuseiso ~/Downloads/my-game.iso my-game/rom`. To unmount use `fusermount -u rom`.

Option 2: You can use the mount command (requires superuser):

To mount use: `sudo mount -o loop ~/Downloads/my-game.iso my-game/rom`, and to unmount use `sudo umount rom`.

**Comment by ruanformigoni** [comment]
Gameimage should detect and .exe files inside the rom directory, and give you an option to choose one to install. After you install the game, the tool also asks you to choose which executable from the installation directory to execute to open the game.

**Comment by Blad3forc3** [comment]
Thanks for the help ruanformigoni :)
Another question. Once the game has been installed and requires another exe to run and configure before the game is actually run how do you do this after the appimage has been created?

**Comment by ruanformigoni** [comment]
No problem :) Hmm, you mean a game that requires a disc to open? Like DRM?

**Comment by Blad3forc3** [comment]
No i am trying to get an old windows 98 game installed which uses an old chipset called 3dfx and a program called nglide to emulate the 3dfx chipset.
The normal procedure is mount the image, run the setup from the cd, install the game, install nglide after, configure nglide (once only) then run the game with the cd automatically mounted

**Comment by ruanformigoni** [comment]
I see is [this](https://www.zeus-software.com/downloads/nglide) nglide? My knowledge on this is limited, but see if this would make sense:

1. Create a rom folder, mount the iso in `rom/my-game`.
1. Create another folder `rom/nglide` and move nglide to it.
1. Install the game, then install nglide (gameimage allows you to install multiple programs).
1. Run nglide before packing the game to configure it (?)

The last part ` run the game with the cd automatically mounted`, does nglide mount it for you? You can drop the iso inside the `build/AppDir/app/wine/drive_c` folder, and tell nglide the path to mount the image is `C:\\my-game.iso`. And after packing it should work with just double clicking it.

**Comment by Blad3forc3** [comment]
hit a problem with the mount part. The cd image is a mixed mode cd, meaning it has a data track and cd audio tracks and is refused being mounted by either fuseiso or the mount option (wrong fs-type). The image can be mounted by cdemu under /dev/sr0. Would a softlink in the rom/my-game work point to /dev/sr0?

**Comment by ruanformigoni** [comment]
It should follow the symlink, if it doesn't let me know and I'll make a quickfix.

**Comment by Blad3forc3** [comment]
I have created a symlink in the game/rom directory which opens the correct cd image when clikcing on it but when running through the appimage it only sees the nglide executable to install not anything on cd image to install

**Comment by ruanformigoni** [comment]
I was able to reproduce the issue. Gitlab-CI is generating the new version `v0.2.9` with this fixed.

**Comment by Blad3forc3** [comment]
I am getting places now :) I can install nglide and use a wineprefix command pointing to the newly created wineprefix in the build folder to configure nglide fine. The next executable to run is the setup off the cd which starts but freezes at 76% and freezes. I am guessing it is trying to do a check from the mounted image using the symlink but can't?

**Comment by ruanformigoni** [comment]
It's a bit harder for me to reproduce since I don't have an image like this. Would copying the files from a previous successful installation work?

**Comment by Blad3forc3** [comment]
Mega link with required files..

https://mega.nz/file/mfJjHbqQ#NIOxuyPM-r6dktuPsdVBaQXSWnK2LWaLM7cYm1wx49o

Steps i have took
1.Mount cd image
2.Create symlink in mygame/rom folder to the mounted image
3.Created nglide folder in mygame/rom folder and placed the nglide setup exe there
4.Run ./GameImage.AppImage --platform=wine --name="" --dir=path to game/rom directory
5.Install nglide first (Option 0)
6.Manually run nglide_config.exe in system32 directory in newly created prefix in the build folder and set screen resolution to what you require
7.Try to install the actual game using Setup.exe from the cd image (Option 16)

Setup freezes at 76%

**Comment by ruanformigoni** [comment]
Hi @Blad3forc3 , I've create a new version ([gameimage.zip](https://github.com/ruanformigoni/gameimage/files/10451422/gameimage.zip)) that allows extra install directories in an environment variable:

```sh
export GIMG_DIR_ROM_EXTRA="/run/media/iso"
```

And then execute gameimage, please check if that works.

**Comment by ruanformigoni** [comment]
@Blad3forc3 Replace the path with the path to your mounted image, in your case `/dev/sr0/`, so:

```sh
export GIMG_DIR_ROM_EXTRA="/dev/sr0/"
```

After this run gameimage normally, and it should find the files inside the extra install directory. Thanks for helping me improve gameimage :smile:

**Comment by Blad3forc3** [comment]
I have tried with the export command but it still gets stuck installing at 76%.
Cdemu actually mounts the image at /media/ade/980310_1103 and when i make a symlink to that in the game/rom directory it does open the right directory on the cd when clicking it.
I tried export GIMG_DIR_ROM_EXTRA=/media/ade/980310_1103 then the gameimage appimage but as i said it still gets stuck installing from the image at 76%

**Comment by ruanformigoni** [comment]
Hi @Blad3forc3 , have you erased the build dir completely after the failed attempts? Remember to erase the symlink, if you managed to make this work before, you can just copy the wineprefix that works to `AppDir/app/wine`. And when you run gameimage again, it'll use the pre-configure prefix.

**Comment by Blad3forc3** [comment]
Yes i did erase and start afresh and deleted the old symlink but it still froze at 76%. Ive tried copying the prefix that works into the AppDir/app/wine dir and run gameimage again and it completes the appimage but wont run as i get this error message from the command line

wine: could not load kernel32.dll, status c0000135

ive tried selecting both win32 and win64 during the gameimage setup but both create this error. I am guessing it has something to with the wine version that is downloaded during the earlier part of gameimage

**Comment by ruanformigoni** [comment]
Getting close :D Could you send me the output of `wine --version` of the version of wine that it works in (outside of the appimage). I can try to create a wine appimage with this version for you to try out.

**Comment by Blad3forc3** [comment]
The wine version that works is wine-1.5.0.
This version was installed with playonlinux 
https://www.playonlinux.com/wine/binaries/phoenicis/upstream-linux-x86/

**Comment by Blad3forc3** [comment]
ANy update on this?

**Comment by ruanformigoni** [comment]
Hey! I'm trying to make a workaround for older wine versions, it is taking some time, sorry about that.

**Comment by Blad3forc3** [comment]
All good! In the meantime i have managed to get this running but the graphics are a little corrupt using any of the wine versions you supply. It definitely is a supplied wine version thing

**Comment by Blad3forc3** [comment]
Another question about this, is it possible to have the image of the game on the c drive within the appimage and automount it?
I was thinking of something like cdemu load 0 "c:\<image>.iso" in the apprun file

**Comment by ruanformigoni** [comment]
Yes, you can copy the image to, let's say: `build/AppDir/app/myimage.img`, then add this to AppRun:

```sh
cdemu load 0 "$APPDIR/app/myimage.img"
```

The `APPDIR` variable refers to the `AppDir` directory.

**Comment by Blad3forc3** [comment]
You, sir, are a God. That works great. I have been having great success with some old windows games and putting them in appimages. This completes the party.

**Comment by Blad3forc3** [comment]
Another question ruanformigoni,
Some of my games are so old they dont run unless i limit the cpu to 1 core with this command
schedtool -a 0x1 -e wine "path to exe"
Where exactly in the apprun file can i replicate this command before wine runs?

**Comment by ruanformigoni** [comment]
In the end of AppRun, from:
```sh
if [ "$@" ]; then
  "$WINE" "$@"
else
  "$WINE" "$exec"
fi
```
to:
```sh
if [ "$@" ]; then
  "$WINE" "$@"
else
  schedtool -a 0x1 -e "$WINE" "$exec"
fi
```

**Comment by Blad3forc3** [comment]
Wonderful! I sometimes use dgvoodoo to upscale the graphics to 1080 in these old games with a command like:
WINEDLLOVERRIDES-d3d8,d3dimm,ddraw=n,b wine "path to exe"
Is it the same principle?

if [ "$@" ]; then
  "$WINE" "$@"
else
  WINEDLLOVERRIDES-d3d8,d3dimm,ddraw=n,b "$WINE" "$exec"
fi

**Comment by ruanformigoni** [comment]
Yes, although I think that this would be clearer to read:

```sh
export WINEDLLOVERRIDES="d3d8,d3dimm,ddraw=n,b"

if [ "$@" ]; then
  "$WINE" "$@"
else
  schedtool -a 0x1 -e "$WINE" "$exec"
fi
```

**Comment by Blad3forc3** [comment]
I dont combine the two normally so as long as i add 

export WINEDLLOVERRIDES="d3d8,d3dimm,ddraw=n,b"

to the apprun file it will work?

**Comment by ruanformigoni** [comment]
Yes, this would also work:
```sh
export WINEDLLOVERRIDES="d3d8,d3dimm,ddraw=n,b"

if [ "$@" ]; then
  "$WINE" "$@"
else
  "$WINE" "$exec"
fi
```

**Comment by Blad3forc3** [comment]
Great thanks!!

**Comment by Blad3forc3** [comment]
Could this be added to the apprun file to check for schedtool?

if ! command -v schedtool &>/dev/null; then
  notify-send "schedtool not found, please install it to use the gameimage"
  exit 1
fi

Ideally these would be incorporated into the appimage if not installed

**Comment by ruanformigoni** [comment]
Yes, you could create appimages of these packages with [arch2appimage](https://github.com/hanzala123/arch2appimage), here is a [video tutorial](https://www.youtube.com/watch?v=Wtnm7VFXWzI) of it. And then use it from the appdir, e.g., `$APPDIR/app/schedtool.AppImage`.

**Comment by Blad3forc3** [comment]
It doesnt seem to like schedtool because it doesnt have a desktop file or image associated with it

**Comment by ruanformigoni** [comment]
You mean something like this?

```sh
if command -v cdemu &>/dev/null; then
  notify-send "Mounting image $APPDIR/app/myimage.img with cdemu..."
  cdemu load 0 "$APPDIR/app/myimage.img"
else
  notify-send "cdemu not found, continuing without mounted image..."
fi
```

**Comment by Blad3forc3** [comment]
ive got an appimage of cdemu created called gcdemu.AppImage.
I was thinking the script could check if cdemu is installed or not on the host system and if not invoke gcdemu.appimage and automatically mounts the iso file then continues to run normally and if the script already finds cdemu installed on the host system it bypasses the gcdemu.appimage and continues to mount the iso file then continues to load as normal

**Comment by ruanformigoni** [comment]
This checks if the host has `cdemu` installed, if it is, use it. Otherwise, it uses the `AppImage`.

```sh
if command -v cdemu &>/dev/null; then
  notify-send "Using host-provided cdemu..."
  cdemu load 0 "$APPDIR/app/myimage.img"
else
  notify-send "Using cdemu AppImage..."
  "$APPDIR/app/cdemu.AppImage" load 0 "$APPDIR/app/myimage.img"
fi
```

**Comment by Blad3forc3** [comment]
Also, in one particular appimage after its created it is complaining about not enough disc space to allocate the virtual heap. I have uploaded the appimage for you

https://mega.nz/file/7GwESB4A#nAfCKYkCrZzXVEawzSUeOL3ngLx8uNuyDkN1ul0G4wI

the game runs fine in its own prefix from my hard drive

**Comment by Blad3forc3** [comment]
It looks like some games need to write files or update config files in the game directory while they are being run. Is there any work around for this?

**Comment by ruanformigoni** [comment]
Hi @Blad3forc3, I have implemented a workaround for this, instructions [here](https://gitlab.com/formigoni/gameimage#where-to-install-the-game). I'll close the issue since the original issue is fixed, however we can keep the discussion going :smile:

**Comment by Blad3forc3** [comment]
THanks for all this ruanformigoni. The appimage i uploaded, I have tried with the new export command
export GIMG_INSTALL_LOC=prefix
but it still gives me a virtual heap error

**Comment by ruanformigoni** [comment]
No problem! What wine dist are you using? Did you completely remove the build dir before rebuild? Could you also provide the output of gameimage? Thanks!

**Comment by Blad3forc3** [comment]
I totally started from scratch and it is saving the settings data and savegames now. Just about to try the one that had the virtual heap problem

**Comment by ruanformigoni** [comment]
There is an alternative that I'm currently working on `GIMG_INSTALL_LOC=unionfs`, this compresses the game and still allows it to write data to its directory. I'll just running some tests before publishing this feature.

**Comment by Blad3forc3** [comment]
Great stuff! The fixes have also made my appimage with the heap error work too

**Comment by Blad3forc3** [comment]
I have re-tested all my appimages with the new version and they all appear to work fine! :dancing_women: 
Being able to compress the config dirs would save alot of space for sure
I will keep testing more as i go...

**Comment by ruanformigoni** [comment]
Version `v0.4.8` is out, you should be able to compress and write the game saves now. This is the default behavior now, no need to export a variable to configure it, more info [here](https://gitlab.com/formigoni/gameimage#how-to-package-the-software).

**Comment by Blad3forc3** [comment]
Thanks. Is there a package i need to install to take advantage of this?

**Comment by ruanformigoni** [comment]
No, it's included

**Comment by Blad3forc3** [comment]
So, As long as i select export GIMG_PKG_TYPE=unionfs the process will automatically compress into 1 file?

**Comment by ruanformigoni** [comment]
Yes, also that is the default behavior now, so no need to export.

**Comment by Blad3forc3** [comment]
i've download the new gameimage and created a new build dir but when i create the appimage and run it i still get a hidden config dir that is as large as the appimage

**Comment by ruanformigoni** [comment]
The contents of the wineprefix are mounted in the hidden config dir when the appimage is running. Check the size of the hidden directory with:

```sh
killall unionfs
du -sh .my-game.AppImage.config
```

**Comment by Blad3forc3** [comment]
ok, the appimage is 1.6GB while running the game and reduces to 2.8mb when the game is closed.
I take it the killall unionfs isnt needed after running an appimage?

**Comment by ruanformigoni** [comment]
It is not needed, it was to double check if unionfs was unmounted, before you calculated the size of the config directory.

**Comment by Blad3forc3** [comment]
i have to manually kill the unionfs before the size goes down in the config file

**Comment by ruanformigoni** [comment]
Then it is a bug, I'll take a look

**Comment by Blad3forc3** [comment]
OK. I was surprised to see a hidden config file created. Is this still needed if you transfer the appimage to another computer?

**Comment by ruanformigoni** [comment]
Only if you want to keep you game saves / data, otherwise, it should still work with only the appimage file. The hidden config dir is required to store this additional data, since the appimage uses a read-only filesystem.

**Comment by Blad3forc3** [comment]
Ah right so any save games/config files changes get stored in this hidden config file and must be transferred with the appimage to another computer but the appimage will be reloaded with the config file/save games stored ok?

**Comment by ruanformigoni** [comment]
Yes, when it is restarted it finds the already existing config dir with modified files.

**Comment by Blad3forc3** [comment]
Cool. Would a simple killall unionfs at the end of the appdir file be enough to automatically kill the process so we dont have a large hidden config file?

**Comment by ruanformigoni** [comment]
I was using `fusermount -u`, which seems to fail sometimes. Now I've replaced with `pkill -f "unionfs.*\$WINEPREFIX"`, I chose to no use `killall` because the user might have another program running that uses unionfs.

**Comment by ruanformigoni** [comment]
Hotfix for unmount released in `v0.4.9`.

**Comment by Blad3forc3** [comment]
Yes, that appears to work fine now. A little niggle though, when you select a name with uppercase letters at the start

./GameImage.AppImage --platform="wine" --name="AbCdE" --dir="blah blah"

the resulting appimage is always lowercase

**Comment by ruanformigoni** [comment]
That is a requirement for some packaging methods.

**Comment by Blad3forc3** [comment]
Oh ok, it just never used to do that. i'll keep testing this new version but all looks good

**Comment by Blad3forc3** [comment]
All good so far. I will write a step by step to add here to elaborate on the main how to then if i have any problems i will create a new issue

**Comment by Blad3forc3** [comment]
How do i add an argument to the exe file within the appimage?

example wine.exe -high.

How do i add the -high?

**Comment by ruanformigoni** [comment]
Hey! I'm working on something for that, will update when it's released.

**Comment by ruanformigoni** [comment]
Feature implemented, more information [here](https://gitlab.com/formigoni/gameimage#launch-command).

**Comment by Blad3forc3** [comment]
"$WINE" my-app.exe -high

Is this in the apprun file? If so, does it replace a command there or just added?

**Comment by ruanformigoni** [comment]
You can replace the default command:

```sh
"$YQ" --exit-status 'tag == "!!map" or tag == "!!seq"' "$YAML" &>/dev/null || echo "cmd: \"{wine} {exec} -high\"" > "$YAML"
```

**Comment by Blad3forc3** [comment]
with this appimage and the new gameinage when i try to run it i get this error

mkdir: cannot create directory ‘/media/ade/Temp/Game Appimages/.Diablo II.AppImage.config/wine/drive_c/users/ade/My Documents’: File exists

**Comment by ruanformigoni** [comment]
Hi @Blad3forc3, thanks a lot for the feedback, this should be now fixed in `v0.5.2`.
